scrapy/addons.py

```diff
@@ -0,0 +1,425 @@
+            comp = self.component or self
+            if self.component_key:
+                # e.g. for DOWNLOAD_HANDLERS: {'http': 'myclass'}
+                k = self.component_key
+                v = comp
+            else:
+                # e.g. for DOWNLOADER_MIDDLEWARES: {'myclass': 100}
+                k = comp
+                v = config.get("order", self.component_order)
+            settings[self.component_type].update({k: v}, "addon")
+
+    def export_basics(self, settings):
+        """Export the :attr:`basic_settings` attribute into the settings object.
+
+        All settings will be exported with ``addon`` priority (see
+        :ref:`topics-api-settings`).
+
+        :param settings: Settings object into which to expose the basic settings
+        :type settings: :class:`~scrapy.settings.Settings`
+        """
+        for setting, value in (self.basic_settings or {}).items():
+            settings.set(setting, value, "addon")
+
+    def export_config(self, config, settings):
+        """Export the add-on configuration, all keys in caps and with
+        :attr:`settings_prefix` or :attr:`name` prepended, into the settings
+        object.
+
+        For example, the add-on configuration ``{'key': 'value'}`` will export
+        the setting ``ADDONNAME_KEY`` with a value of ``value``. All settings
+        will be exported with ``addon`` priority (see
+        :ref:`topics-api-settings`).
+
+        :param config: Add-on configuration to be exposed
+        :type config: ``dict``
+
+        :param settings: Settings object into which to export the configuration
+        :type settings: :class:`~scrapy.settings.Settings`
+        """
+        if self.settings_prefix is False:
+            return
+        conf = self.default_config or {}
+        conf.update(config)
+        prefix = self.settings_prefix or self.name
+        # Since default exported config is case-insensitive (everything will be
+        # uppercased), make mapped config case-insensitive as well
+        conf_mapping = {k.lower(): v for k, v in (self.config_mapping or {}).items()}
+        for key, val in conf.items():
+            if key.lower() in conf_mapping:
+                key = conf_mapping[key.lower()]
+            else:
+                key = (prefix + "_" + key).upper()
+            settings.set(key, val, "addon")
+
+    def update_settings(self, config, settings):
+        """Export both the basic settings and the add-on configuration. I.e.,
+        call :meth:`export_basics` and :meth:`export_config`.
+
+        For more advanced add-ons, you may want to override this callback.
+
+        :param config: Add-on configuration
+        :type config: ``dict``
+
+        :param settings: Crawler settings object
+        :type settings: :class:`~scrapy.settings.Settings`
+        """
+        self.export_component(config, settings)
+        self.export_basics(settings)
+        self.export_config(config, settings)
+
+
+class AddonManager(Mapping):
```

Comment:
If we proceed with removal of various features, we should check which code in the AddonManager is still valuable.